src/EditorFeatures/Core.Wpf/Utilities/RoslynServiceExtensions.cs

```diff
@@ -0,0 +1,186 @@
+        /// </summary>
+        public static TServiceType GetServiceOnMainThread<TServiceType>(this IServiceProvider serviceProvider) where TServiceType : class
+            => serviceProvider.GetServiceOnMainThread<TServiceType, TServiceType>();
+
+        /// <summary>
+        /// Gets a service interface from a service provider.
+        /// </summary>
+        /// <typeparam name="TService">The service type</typeparam>
+        /// <typeparam name="TInterface">The interface type</typeparam>
+        /// <param name="serviceProvider">The service provider</param>
+        /// <returns>The requested service interface. Never <see langword="null"/>.</returns>
+        /// <exception cref="ServiceUnavailableException">
+        /// Either the service could not be acquired, or the service does not support
+        /// the requested interface.
+        /// </exception>
+        public static TInterface GetService<TService, TInterface>(
+            this IServiceProvider serviceProvider,
+            JoinableTaskFactory joinableTaskFactory)
+            where TInterface : class
+        {
+            return GetService<TService, TInterface>(serviceProvider, joinableTaskFactory, throwOnFailure: true)!;
+        }
+
+        /// <summary>
+        /// Gets a service interface from a service provider.
+        /// </summary>
+        /// <typeparam name="TService">The service type</typeparam>
+        /// <typeparam name="TInterface">The interface type</typeparam>
+        /// <param name="serviceProvider">The service provider</param>
+        /// <param name="throwOnFailure">
+        /// Determines how a failure to get the requested service interface is handled. If <see langword="true"/>, an
+        /// exception is thrown; if <see langword="false"/>, <see langword="null"/> is returned.
+        /// </param>
+        /// <returns>The requested service interface, if it could be obtained; otherwise <see langword="null"/> if
+        /// <paramref name="throwOnFailure"/> is <see langword="false"/>.</returns>
+        /// <exception cref="ServiceUnavailableException">
+        /// Either the service could not be acquired, or the service does not support
+        /// the requested interface.
+        /// </exception>
+        [SuppressMessage("Usage", "VSTHRD102:Implement internal logic asynchronously", Justification = "Intentional to match required semantics for IServiceProvider.GetService")]
+        public static TInterface? GetService<TService, TInterface>(
+            this IServiceProvider serviceProvider,
+            JoinableTaskFactory joinableTaskFactory,
+            bool throwOnFailure)
+            where TInterface : class
+        {
+            Requires.NotNull(serviceProvider, nameof(serviceProvider));
+
+            return joinableTaskFactory.Run(async () =>
+            {
+                await joinableTaskFactory.SwitchToMainThreadAsync();
+
+                var service = serviceProvider.GetService(typeof(TService));
```

Comment:
should this just call: GetServiceOnMainThread at this point?

Reply:
No. The error handling behavior is different.

`GetServiceOnMainThread` is a Roslyn-specific behavior. The IDE implementation of these `GetService` calls always use JTF.